### PR TITLE
Annotate Binder.cs and BuckStopsHereBinder

### PR DESCRIPTION
**Pull request by 333fred**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 1]
```diff
@@ -697,12 +716,18 @@ internal BoundExpression WrapWithVariablesIfAny(CSharpSyntaxNode scopeDesignator
             var locals = this.GetDeclaredLocalsForScope(scopeDesignator);
             return (locals.IsEmpty)
                 ? expression
-                : new BoundSequence(scopeDesignator, locals, ImmutableArray<BoundExpression>.Empty, expression, expression.Type) { WasCompilerGenerated = true };
+                : new BoundSequence(scopeDesignator, locals, ImmutableArray<BoundExpression>.Empty, expression, getType()) { WasCompilerGenerated = true };
+
+            TypeSymbol getType()
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: >getType [](start = 23, length = 7)

ah, the irony, we're making a `BoundSequence` here ;-P

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 2]
```diff
@@ -234,6 +234,7 @@ internal bool AreNullableAnnotationsEnabled(SyntaxTree syntaxTree, int position)
 
         internal bool AreNullableAnnotationsEnabled(SyntaxToken token)
         {
+            if (token.SyntaxTree is null) return false;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: can we cover this?

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 3]
```diff
@@ -28,28 +28,29 @@ internal partial class Binder
         /// </summary>
         internal Binder(CSharpCompilation compilation)
         {
-            Debug.Assert(compilation != null);
+            RoslynDebug.Assert(compilation != null);
+            RoslynDebug.Assert(this is BuckStopsHereBinder);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: This seems like an odd requirement to put in a constructor.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder.cs]: It's the only thing allowed to have a null _next.

**Comment by RikkiGibson** [comment]
This passes after merging in latest master locally, so will merge on @333fred's behalf

**Comment by 333fred** [comment]
Thanks Rikki!

**Pull request merged**